### PR TITLE
Darling MCP declares its peer stores instead of answering "unknown server" (#2339)

### LINKED ISSUE #2339: Multi-store fleets: each Darling MCP server is silent about its siblings — an agent cannot know which box answers for which servers
Raised by Erik while reviewing the #2296 split (and about to get worse with a third, PostgreSQL-only box):

> i'm also interested in how the darling mcp handles having two destinations on each box now. how do agents know which to use?

## Today's behavior

Each box's MCP server exposes the same ~101 tool names over ITS store only. `list_servers` answers with that store's registry; a server monitored by a sibling box simply resolves as not-found, indistinguishable from a server nobody monitors. Nothing in the instructions, `list_servers`, or any error names the split — so an agent connected to the use2 endpoint asking about `prod-sql-use1-beta-01` gets 'unknown server' when the true answer is 'the use1 box has that one.' With the fleet split by design (primaries in-region on use1, replicas in-region on use2, PostgreSQL heading to a third box), that's now the normal case, not an edge.

## Fix shape, two tiers

1. **Cheap, ships first — declared peers.** A `peers` block in config (name + what-it-covers, no credentials): the MCP instructions gain a paragraph ('THIS store monitors the 42 use1 primaries; prod-sql-use2-monitor-01 monitors their readable replicas; …'), `list_servers` gains a `peer_fleets` summary, and the server-resolution miss message changes from 'unknown server' to 'not monitored HERE — matches the declared coverage of <peer>.' No cross-store connectivity, purely disclosure — an agent picks the right endpoint or tells its human which one to point at.
2. **Later, maybe — federation.** Actual cross-store reads (one endpoint proxying to peers). Much bigger surface (auth between stores, latency, partial failures) and possibly never worth it if tier 1 makes the split legible.

Tier 1 also documents itself into the census pin naturally (the instructions text is already inventory-pinned).

### PR DESCRIPTION
**Pull request by erikdarlingdata**
Closes #2339 — **tier 1 only** (declared peers / disclosure). No federation, no cross-store connectivity, and none is built here.

## The problem

The fleet is now split across several Darling boxes, one store each: the use1 box monitors 42 SQL Server primaries, the use2 box monitors their readable replicas, a third box is heading for PostgreSQL. Each box's MCP server answers over **its** store only, and nothing in the instructions, `list_servers`, or any error named the split — so an agent connected to the use2 endpoint asking about `prod-sql-use1-beta-01` got `Could not resolve server`, which is indistinguishable from *"nobody monitors that server."* With the fleet split by design, that is now the normal case rather than an edge.

## What this adds

An optional `peers` block in `darling.json` — `thisStoreCovers`, plus per-peer `name` / `covers` / optional `matches` name-substrings — disclosed at the places an agent forms its picture of the fleet:

1. **The MCP instructions** gain a *Fleet Coverage* section, placed between the read-only preamble and the tool census so an agent learns **which** store it is talking to before it starts planning against the tool list.
2. **`list_servers`** gains `this_store_covers`, a `peer_fleets` array, and a `peer_note` — including on the empty-registry answer, which is prose rather than JSON and would otherwise be the one path where the declaration silently vanished (a store with nothing registered is a fresh or just-restarted box, the worst place to drop it).
3. **The server-resolution miss** appends the disclosure to the existing available-servers listing.

The miss now reads (verbatim from the harness):

```
Could not resolve server. Available servers:
prod-sql-use1-beta-01

'prod-sql-use2-beta-01' is not monitored HERE, and it matches the declared coverage of peer
store prod-sql-use2-monitor-01 — the readable replicas of those same 42 primaries, from
us-east-2. That is a SEPARATE Darling store with its own MCP endpoint; this server cannot
read it, so point the client at that endpoint (or tell your operator which store answers for
this server) rather than concluding the server is unmonitored. This store covers: the 42
us-east-1 SQL Server primaries.
```

It is **appended**, not substituted: `Could not resolve server.` is the prefix callers (and several tests) key off, and the local listing is still the right answer to the commonest miss, a typo.

## Decisions worth reviewing

**Disclosure only, and it says so everywhere.** No address and no credential in the block, and nothing behind it — the service never contacts a peer, cannot read one's data, and cannot tell whether one is running. Every message states that, because an agent told a sibling exists will otherwise try to route a query at it.

**A peers block that fails validation is refused whole.** All of this text is sent verbatim to every connected MCP client, so `PeersConfig.Validate` rejects `password=` / `connectionString` / `Integrated Security=` etc. in any peer string, and a nameless peer (an agent told "some other store has it", with no name to point a human at, is no better off than before). The guard runs **inside `DarlingPeerDirectory.Publish`**, not only in config validation — see the review round below for why that distinction turned out to be load-bearing. A refusal publishes `Snapshot.Empty`: the whole block, not the valid subset, because an unfinished block asserting coverage that may be wrong is worse than none. A peer with a name and no `covers` is fine and renders as just the name.

**`matches` is plain case-insensitive substrings** — no globbing, no regex. It exists to answer "which region/role prefix is this name?", and a pattern language is a config surface with its own failure modes. Blank entries are dropped: an empty substring matches every name and would make one peer claim the whole fleet. A peer that declares none is still disclosed everywhere, it just cannot be singled out on a miss — and the miss message says that rather than implying the server is unmonitored.

**An empty `peer_fleets` carries its own note.** It means *either* this is the only Darling store *or* the operator never declared the siblings, and the service cannot tell those apart — so `peer_note` says exactly that instead of letting an empty array read as "you are looking at the whole fleet."

**The snapshot is ambient, not injected.** ~90 tool methods resolve a server name through `DarlingServerResolver` taking only an `NpgsqlDataSource`; threading a parameter through all of them plus the web dashboard's read dispatch would touch every one to deliver a constant. It is an immutable record behind a `volatile` field, published from both the worker and the MCP host (either may reach its config first; they read the same file so they cannot disagree). Every pure function also has an explicit-snapshot overload, which is what the tests use.

**Nothing declared = nothing changes, with one named exception.** `Snapshot.Empty` is the default, so the three *prose* surfaces are byte-for-byte what they were: the instructions (`Build(Empty)` pinned by reference equality with `Text`), the resolution miss (pinned by exact string equality), and `list_servers`' empty-registry sentence.

The exception, raised in review and now called out in the CHANGELOG, the test class doc and the README rather than folded into an "unchanged" claim: **`list_servers`' JSON envelope carries `this_store_covers` / `peer_fleets` / `peer_note` on every response**, declared or not — so a client comparing that tool's exact shape sees three new keys on upgrade even if it never writes a `peers` block. That is the design, not an oversight: a note that appeared only once peers were declared would say nothing in precisely the case that produces the wrong conclusion.

**Lite gets no twin** — it has no peers concept and no central store — and nothing in `PerformanceMonitor.Collectors` / `PerformanceMonitor.Common` was touched.

## Review round

Three findings from `claude[bot]`, all reproduced before fixing, all answered on their threads.

1. **The credential guard never covered the path that broadcasts.** `PeersConfig.Validate` ran only from `DarlingConfig.Validate`, which only the worker calls — and the worker's abort is a `return` from its own hosted service, not a process exit (verified: no `StopApplication` or `Environment.Exit` anywhere in `Program.cs`). The MCP host loads its own copy of `darling.json` and deliberately never validates it, because its fail-closed checks are host-local by design. So a credential pasted into a peer description stopped collection and then went out to every connected client — and with the aborted worker never reaching its own publish, the host's unvalidated publish became the *only* value in the ambient snapshot. Fixed at the category rather than the call site: `Publish` now validates and refuses, returning a `PublishResult` record struct carrying the problems so a caller cannot take the snapshot and drop the reason. The ambient snapshot can only be written through `Publish`, so a future third publish site cannot reintroduce the hole.
2. **An explicit `"stores": null` skipped the `thisStoreCovers` guard.** `System.Text.Json` assigns null over the property initializer (an omitted key leaves the default), and the self-text check sat after the per-peer loop behind an early return. Reproduced — `Validate() problems: 0` with the credential reaching the rendered disclosure — then fixed by **ordering** rather than a second check: an unconditional guard cannot be bypassed by a shape nobody enumerated. My own tests missed it because they all built `PeersConfig` in C#, where the initializer guarantees a non-null list.
3. **The "byte-for-byte unchanged" claim overstated the `list_servers` case** — corrected above and in three files; documentation only, no behavior change, and the pin that exposed the discrepancy was left exactly as it was.

Also folded in: the multi-claimant resolution miss now agrees in number. Two peers can legitimately both claim a name through overlapping `matches`, and the follow-on sentence said "That is a SEPARATE Darling store" about a list of two.

## Verification

`PerformanceMonitor.Darling.Service`, `Darling.Tests` and `Lite.Tests` all build clean (`EnableWindowsTargeting`, 0 warnings). The Windows-only suites **cannot run on macOS**, so CI is the arbiter for the 23 new xUnit facts in `Darling/Darling.Tests/DarlingPeerDisclosureTests.cs`.

To not defer the logic to CI, a throwaway `net10.0` console harness ran **114 assertions** against the real build: config parse/validate (including the shipped sample), normalization and matching, the instructions section and its placement, the rendered `list_servers` JSON in both the declared and empty cases, the empty-registry prose, all five resolution-miss shapes, the ambient publish/refuse/reset, and both code findings — including that leaked text reaches neither the rendered instructions nor a resolution miss, and that the *valid* peer in a refused block is not disclosed either. **Negative control:** neutering `ResolutionMissDisclosure` turns 9 of them red, so they test the fix rather than passing by accident.

Two specific pins that could have bitten silently:

- The **cross-app census pin** (`Lite.Tests/CrossAppMcpToolInventoryPinTests.DarlingInstructionsCensus_MatchesTheScannedInventory`) parses the tool-census sentence out of `DarlingMcpInstructions.cs`. Splitting the const into `Preamble`/`Body` leaves that sentence intact; the harness re-runs the pin's own regex against the shipped string and checks its arithmetic.
- `darling.sample.json` is the **real fixture** for `DarlingNetworkConfigEditor`'s comment-preserving text surgery, and adding a `peers` block moved which root member is last. The harness re-ran every upsert/remove against the edited sample (all three network parents, plus the new one) and confirmed each result still parses with the peers block intact.

Not verified locally: the xUnit suites themselves, and no live-store or on-box behaviour — the disclosure is pure over config, so there is no collector or store path to exercise.

One thing deliberately **not** changed: that the worker's `Validate()` abort does not stop the process is pre-existing for every config problem (`servers must contain at least one entry` behaves the same way), so it is a separate question from this PR.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: Darling/PerformanceMonitor.Darling.Service/Mcp/DarlingMcpHostService.cs [hunk 1]
```diff
@@ -445,6 +445,21 @@ listen value is itself loopback or a wildcard (0.0.0.0/::), which would collide
             builder.Services.AddSingleton<NpgsqlDataSource>(postgres);
             builder.Services.AddSingleton(new DarlingAnalysisService(postgres, planFetcher, _logger));
 
+            /* #2339: publish the declared peer stores before the instructions are rendered, so the same
+               snapshot feeds the instructions section, list_servers' peer_fleets block, and the
+               server-resolution miss message. Publishing here as well as in the worker is deliberate: either
+               may reach its config first, the value is identical (both read darling.json), and the disclosure
+               should not depend on which one won. An empty declaration is Snapshot.Empty, which leaves every
+               one of those three surfaces exactly as it was. */
+            var declaredPeers = DarlingPeerDirectory.Publish(config.Peers);
```
[the remark this answers is not in the document]
Comment (erikdarlingdata) [reply Darling/PerformanceMonitor.Darling.Service/Mcp/DarlingMcpHostService.cs]: Confirmed and fixed in e35766d2. I verified both halves of the premise before touching anything: there is no `StopApplication` or `Environment.Exit` anywhere in `Program.cs`, so the worker's `return` on a `Validate()` problem ends only its own hosted service, and this host does load and publish independently. Your point that the aborted worker never reaches its own `Publish` — making this host's unvalidated publish the *only* value in `DarlingPeerDirectory.Current` — is exactly right, and it's the part I'd missed.

I took it one level up from the suggestion, because this is the second instance of one category (the guard not covering a path that nobody enumerated). Rather than adding the check to this call site, `DarlingPeerDirectory.Publish` now validates and refuses:

```csharp
internal static PublishResult Publish(PeersConfig? config)
{
    var problems = PeersConfig.Validate(config);
    if (problems.Count > 0)
    {
        s_current = Snapshot.Empty;
        return new PublishResult(Snapshot.Empty, problems);
    }
    ...
}
```

`s_current` can only be written through `Publish`, so the guard is now structural rather than something each publish site has to remember — a future third one cannot reintroduce the hole. `PublishResult` is a record struct carrying the problems so a caller cannot take the snapshot and drop the reason; both call sites log them at Critical.

A refusal publishes `Snapshot.Empty` — the **whole** block, not the valid subset. A peers block that failed validation is one the operator hasn't finished, and half a disclosure would assert coverage that may be wrong while the log says the config is broken. Losing disclosure for one restart is recoverable; leaking a credential isn't.

Harness (Windows suites can't run on macOS) reproduced the leak first, then confirmed the leaking text reaches neither the rendered instructions nor a resolution miss, and that the *valid* peer in a refused block isn't disclosed either. Pinned by `Publish_RefusesTheWholeBlockOnAnyValidationProblem`.

One thing I deliberately did **not** change: that the worker's `Validate()` abort doesn't stop the process is pre-existing for every config problem (`servers must contain at least one entry` behaves the same way — collection stops while MCP keeps serving a store that will never fill). That's arguably fine given each host self-gates, but it's a separate question from this PR and I didn't want to widen the diff into it.

File: Darling/PerformanceMonitor.Darling.Service/DarlingConfig.cs [hunk 2]
```diff
@@ -1232,3 +1245,153 @@ public sealed class MonitoredServer
...
+    /// A peer with a name but no <c>covers</c> sentence is allowed — half a disclosure still names an
+    /// endpoint — and renders as just the name.
+    /// </summary>
+    public static IReadOnlyList<string> Validate(PeersConfig? peers)
+    {
+        var problems = new List<string>();
+        if (peers?.Stores is null)
+        {
+            return problems;
+        }
```
[the remark this answers is not in the document]
Comment (erikdarlingdata) [reply Darling/PerformanceMonitor.Darling.Service/DarlingConfig.cs:1327]: Real, reproduced, and fixed in e35766d2. Your exact config, run against the build before the fix:

```
  Stores is null after "stores": null  -> True
  Validate() problems: 0 []
  LEAK REACHES DISCLOSURE: internal db, see Password=hunter2
```

So `System.Text.Json` does assign null over the property initializer for an explicit `"stores": null`, the early return did skip the self-text guard, and the credential did reach `FromConfig(...).ThisStoreCovers` — from where all three surfaces disclose it.

You also called the cause correctly: none of my tests exercised `Stores == null`, because they all built `PeersConfig` in C# where the initializer guarantees a non-null list. The JSON path was the only way in and I never sent it a null.

Fixed by **ordering, not an extra check** — the `thisStoreCovers` guard now runs first, before anything can short-circuit, with the `Stores is null` return after it. An unconditional guard can't be bypassed by a shape nobody thought to enumerate, whereas a second copy of the check would just be another thing to keep in sync. The null guard also relaxed from `peers?.Stores is null` to `peers is null` so a null `Stores` no longer skips the whole function.

Belt and braces on top: `DarlingPeerDirectory.Publish` now runs `PeersConfig.Validate` itself and refuses (see the other thread), so even if a future guard is bypassed again, nothing invalid reaches the ambient snapshot that the disclosure surfaces read.

Pinned by `Validate_ChecksThisStoreCovers_EvenWhenStoresIsExplicitJsonNull`, which parses your config verbatim, asserts `Stores` is genuinely null (so the test still exercises the real shape rather than a fixture that can't reach it), and asserts nothing is published.

File: Darling/PerformanceMonitor.Darling.Service/Mcp/DarlingMcpHostService.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (erikdarlingdata) [reply Darling/PerformanceMonitor.Darling.Service/Mcp/DarlingMcpHostService.cs]: Same finding as the thread above on this file — answered there in full. Short version: confirmed (no `StopApplication`/`Environment.Exit` in `Program.cs`, so the worker's abort ends only the worker), and fixed in e35766d2 by moving the guard into `DarlingPeerDirectory.Publish` rather than into this call site, so the ambient snapshot cannot be written without validation from any publish site, present or future. Your note that it combines with the `Stores == null` bypass to leave the guard with essentially no enforcement on the broadcast path was accurate — both are fixed and pinned.

File: Darling/PerformanceMonitor.Darling.Service/Mcp/DarlingMcpDataTools.cs [hunk 3]
```diff
@@ -720,32 +720,66 @@ public static async Task<string> ListServers(
...
+                name = p.Name,
+                covers = p.Covers,
+                matches = p.Matches
+            }),
+            peer_note = peers.Peers.Count == 0
+                ? DarlingPeerDirectory.NoPeersDeclaredNote
+                : DarlingPeerDirectory.PeersDeclaredNote,
+            servers = result
+        }, McpHelpers.JsonOptions);
+    }
```
[the remark this answers is not in the document]
Comment (erikdarlingdata) [reply Darling/PerformanceMonitor.Darling.Service/Mcp/DarlingMcpDataTools.cs:781]: Fair, and fixed as a documentation correction rather than a behavior change — the claim was wrong, not the code.

You're right on all of it: three surfaces are byte-for-byte unchanged (instructions via reference equality, resolution miss via exact string equality, and `list_servers`' empty-registry sentence), and `list_servers`' JSON envelope is the one that isn't. Three new keys plus a multi-sentence `peer_note` land on every response, including the default the sample now ships.

The shape stays as it is, because it's the design and not an oversight: an empty `peer_fleets` means *either* "this is the only Darling store" *or* "the operator never declared the siblings", and this server can't tell those apart. A `peer_note` that only appeared once peers were declared would say nothing in precisely the case that produces the wrong conclusion — which is the conclusion #2339 was filed about. So the honest-empty note has to be unconditional to be worth having.

But an overstated claim is worse than an admitted exception, so the exception is now named in three places instead of being folded into "unchanged":

- **CHANGELOG** — "the PROSE surfaces are byte-for-byte unchanged ... with ONE deliberate exception: `list_servers`' JSON envelope carries `this_store_covers` / `peer_fleets` / `peer_note` on every response ... so a client comparing that tool's exact shape sees three new keys on upgrade."
- **`DarlingPeerDisclosureTests` class doc** — the negative-invariant paragraph now enumerates which three surfaces it covers, with a second paragraph for the exception and why a conditional block would be worse.
- **`Darling/README.md`** — a "Declaring nothing changes nothing, with one exception worth knowing about on upgrade" note in the new `### peers` section, since that's where an operator upgrading will look.

`ListServers_EmptyPeerFleets_SaysWhatItDoesNotProve` is left exactly as it was. It was asserting the right thing all along; the prose around it was what disagreed with it, and a pin that contradicts the docs is the pin being right.

In c02cf1c2 (docs only, no behavior change).

**Comment by erikdarlingdata** [comment]
Reviewed the diff, not the report — merging. I checked the three things that could have made this dangerous rather than useful:

1. **The census pin survived.** The instructions still read `This server exposes 101 tools. 76 are the same names`, so `CrossAppMcpToolInventoryPinTests` still parses it. Inserting a Fleet Coverage section above the census was the risky placement and it held.
2. **It really is disclosure-only.** Grepping `DarlingPeerDirectory` for address/credential/endpoint/connection turns up nothing but doc comments explaining their absence. Tier 1 stayed tier 1.
3. **The default install is untouched.** `disclosure.Length == 0 ? message : ...` means a store declaring no peers produces a byte-identical miss message. That was my main regression worry — a fleet-split feature that changes what every single-store operator sees — and it is explicitly guarded and pinned.

The credential-guard bug you found and fixed mid-review is the good catch here: `DarlingConfig.Validate` is run by the worker, whose failure path is a `return` rather than a process exit, and the MCP host loads its own config and never validates — so the one path that actually broadcasts peer text to clients was the one path the guard never covered. Moving it inside `Publish` is the right shape, since that is where the broadcast is.

Also noted: you corrected your own "all four surfaces byte-for-byte unchanged" claim rather than leaving it standing when `list_servers` always carries the three new keys. That distinction matters to anyone parsing the envelope.
